File: ios/FastVideoAdaptiveRuntime.swift
import Foundation
import Network

@MainActor
internal final class FastVideoAdaptiveRuntime {
  static let shared = FastVideoAdaptiveRuntime()

  private let monitor = NWPathMonitor()
  private let queue = DispatchQueue(label: "com.vivekjm.fastvideo.network")
  private var networkClass = 0
  private var activePlayers = 0
  private(set) var mode = "balanced"

  private init() {
    monitor.pathUpdateHandler = { [weak self] path in
      let value: Int
      if path.status != .satisfied {
        value = 0
      } else if path.usesInterfaceType(.wiredEthernet) {
        value = 4
      } else if path.usesInterfaceType(.wifi) {
        value = 3
      } else if path.usesInterfaceType(.cellular) {
        value = 2
      } else {
        value = 1
      }
      Task { @MainActor in self?.networkClass = value }
    }
    monitor.start(queue: queue)
  }

  func configure(mode: String?) {
    switch mode {
    case "off", "conservative", "aggressive": self.mode = mode ?? "balanced"
    default: self.mode = "balanced"
    }
  }

  func acquiredPlayer() { activePlayers += 1 }
  func releasedPlayer() { activePlayers = max(0, activePlayers - 1) }

  func decision(
    core: RNFVFastCoreBridge,
    bandwidth: Double,
    rebufferRatio: Double,
    droppedFrameRatio: Double,
    width: Int,
    height: Int
  ) -> [String: Any] {
    guard mode != "off" else { return [:] }
    var base = core.adaptiveDecision(
      withBandwidth: max(0, bandwidth),
      rebufferRatio: max(0, rebufferRatio),
      droppedFrameRatio: max(0, droppedFrameRatio),
      activePlayers: max(1, activePlayers),
      width: max(0, width),
      height: max(0, height),
      networkClass: networkClass,
      thermalClass: thermalClass,
      lowPowerMode: ProcessInfo.processInfo.isLowPowerModeEnabled
    ) as? [String: Any] ?? [:]
    if let bitrate = (base["maxBitrateBps"] as? NSNumber)?.doubleValue {
      let multiplier = mode == "conservative" ? 0.78 : mode == "aggressive" ? 1.12 : 1
      base["maxBitrateBps"] = Int64(bitrate * multiplier)
    }
    base["mode"] = mode
    return base
  }

  func stats() -> [String: Any] {
    [
      "adaptivePlaybackEnabled": mode != "off",
      "adaptiveMode": mode,
      "activePlayers": activePlayers,
      "networkClass": networkName,
      "thermalClass": thermalClass,
      "lowPowerMode": ProcessInfo.processInfo.isLowPowerModeEnabled
    ]
  }

  private var networkName: String {
    switch networkClass {
    case 4: return "ethernet"
    case 3: return "wifi"
    case 2: return "cellular"
    case 1: return "constrained"
    default: return "offline"
    }
  }

  private var thermalClass: Int {
    switch ProcessInfo.processInfo.thermalState {
    case .nominal: return 0
    case .fair: return 1
    case .serious: return 2
    case .critical: return 3
    @unknown default: return 1
    }
  }
}

File: android/src/main/java/com/vivekjm/fastvideo/FastVideoAdaptiveRuntime.kt
package com.vivekjm.fastvideo

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.PowerManager
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicInteger

internal object FastVideoAdaptiveRuntime {
  private val activePlayers = AtomicInteger(0)
  @Volatile private var adaptiveMode = "balanced"

  fun configure(value: String?) {
    adaptiveMode = when (value) {
      "off", "conservative", "aggressive" -> value
      else -> "balanced"
    }
  }

  fun onPlayerAcquired() {
    activePlayers.incrementAndGet()
  }

  fun onPlayerReleased() {
    activePlayers.updateAndGet { current -> (current - 1).coerceAtLeast(0) }
  }

  fun activePlayerCount(): Int = activePlayers.get().coerceAtLeast(1)

  fun decision(
    context: Context,
    bandwidthEstimateBps: Long,
    rebufferRatio: Double,
    droppedFrameRatio: Double,
    width: Int,
    height: Int
  ): Map<String, Any?> {
    if (adaptiveMode == "off") return emptyMap()
    val json = FastCoreNative.adaptiveDecision(
      bandwidthEstimateBps.coerceAtLeast(0L).toDouble(),
      rebufferRatio.coerceAtLeast(0.0),
      droppedFrameRatio.coerceAtLeast(0.0),
      activePlayerCount(),
      width.coerceAtLeast(0),
      height.coerceAtLeast(0),
      networkClass(context),
      thermalClass(context),
      isLowPowerMode(context)
    )
    val base = runCatching {
      val value = JSONObject(json)
      value.keys().asSequence().associateWith { key -> value.opt(key).takeUnless { it === JSONObject.NULL } }
    }.getOrDefault(emptyMap())
    if (base.isEmpty()) return base
    val multiplier = when (adaptiveMode) {
      "conservative" -> 0.78
      "aggressive" -> 1.12
      else -> 1.0
    }
    val bitrate = (base["maxBitrateBps"] as? Number)?.toLong()?.let { (it * multiplier).toLong() }
    return if (bitrate != null) base + mapOf("maxBitrateBps" to bitrate, "mode" to adaptiveMode) else base
  }

  fun runtimeStats(context: Context): Map<String, Any> = mapOf(
    "adaptivePlaybackEnabled" to (adaptiveMode != "off"),
    "adaptiveMode" to adaptiveMode,
    "activePlayers" to activePlayers.get(),
    "networkClass" to networkClassName(networkClass(context)),
    "thermalClass" to thermalClass(context),
    "lowPowerMode" to isLowPowerMode(context)
  )

  private fun networkClass(context: Context): Int {
    val manager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager ?: return 0
    val capabilities = manager.getNetworkCapabilities(manager.activeNetwork) ?: return 0
    return when {
      capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED) &&
        capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> 4
      capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> 3
      capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> 2
      capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) -> 1
      else -> 0
    }
  }

  private fun networkClassName(value: Int): String = when (value) {
    4 -> "ethernet"
    3 -> "wifi"
    2 -> "cellular"
    1 -> "constrained"
    else -> "offline"
  }

  private fun thermalClass(context: Context): Int {
    if (Build.VERSION.SDK_INT < 29) return 0
    val power = context.getSystemService(Context.POWER_SERVICE) as? PowerManager ?: return 0
    return when (power.currentThermalStatus) {
      PowerManager.THERMAL_STATUS_CRITICAL, PowerManager.THERMAL_STATUS_EMERGENCY, PowerManager.THERMAL_STATUS_SHUTDOWN -> 3
      PowerManager.THERMAL_STATUS_SEVERE -> 2
      PowerManager.THERMAL_STATUS_MODERATE -> 1
      else -> 0
    }
  }

  private fun isLowPowerMode(context: Context): Boolean {
    val power = context.getSystemService(Context.POWER_SERVICE) as? PowerManager ?: return false
    return power.isPowerSaveMode
  }
}
